src/lib/cost_model/cost_model_logical.cpp

```diff
@@ -1,45 +1,63 @@
 #include "cost_model_logical.hpp"
 
-#include "abstract_cost_feature_proxy.hpp"
-#include "operators/abstract_operator.hpp"
+#include "expression/abstract_expression.hpp"
+#include "expression/expression_utils.hpp"
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "logical_query_plan/join_node.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/union_node.hpp"
+#include "statistics/table_statistics.hpp"
+#include "utils/assert.hpp"
 
 namespace opossum {
 
-std::string CostModelLogical::name() const { return "CostModelLogical"; }
+Cost CostModelLogical::_estimate_node_cost(const std::shared_ptr<AbstractLQPNode>& node) const {
+  const auto output_row_count = node->get_statistics()->row_count();
+  const auto left_row_count = node->left_input() ? node->left_input()->get_statistics()->row_count() : 0.0f;
+  const auto right_row_count = node->right_input() ? node->right_input()->get_statistics()->row_count() : 0.0f;
 
-Cost CostModelLogical::get_reference_operator_cost(const std::shared_ptr<AbstractOperator>& op) const {
-  return estimate_operator_cost(op);
-}
-
-Cost CostModelLogical::_cost_model_impl(const OperatorType operator_type,
-                                        const AbstractCostFeatureProxy& feature_proxy) const {
-  switch (operator_type) {
-    case OperatorType::JoinHash:
-      return feature_proxy.extract_feature(CostFeature::LeftInputRowCount).scalar() +
-             feature_proxy.extract_feature(CostFeature::RightInputRowCount).scalar();
-
-    case OperatorType::TableScan:
-      return feature_proxy.extract_feature(CostFeature::LeftInputRowCount).scalar();
+  switch (node->type) {
+    case LQPNodeType::Join:
+      return left_row_count + right_row_count + output_row_count;
```

Comment:
Your proposed formula taxes cross joins the same as predicated joins. Mine taxes cross joins higher, since output_row_count will be higher for them.
A logical cost model returns the approximate number of tuple read/writes.